signac/contrib/filterparse.py

```diff
@@ -218,14 +241,68 @@ def parse_filter_arg(args, file=sys.stderr):
         if _is_json(args[0]):
             return _parse_json(args[0])
         else:
-            return _with_message(_parse_simple(args[0]), file)
+            key, value = _parse_single(args[0])
+            return _with_message({key: value}, file)
     else:
-        q = {}
-        for i in range(0, len(args), 2):
-            key = args[i]
-            if i + 1 < len(args):
-                value = args[i + 1]
-            else:
-                value = None
-            q.update(_parse_simple(key, value))
+        q = dict(parse_simple(args))
+
         return _with_message(q, file)
+
+
+def _add_prefix(prefix, filter):
+    if filter:
+        for key, value in filter.items():
+            if key in ("$and", "$or"):
+                if isinstance(value, list) or isinstance(value, tuple):
+                    yield key, [dict(_add_prefix(prefix, item)) for item in value]
+                else:
+                    raise ValueError(
+                        "The argument to a logical operator must be a sequence (e.g. a list)!"
+                    )
+            elif "." in key and key.split(".", 1)[0] in ("sp", "doc"):
+                yield key, value
+            elif key in ("sp", "doc"):
+                yield key, value
+            else:
+                yield prefix + key, value
+
+
+def _root_keys(filter):
+    for key, value in filter.items():
+        if key in ("$and", "$or"):
+            assert isinstance(value, (list, tuple))
+            for item in value:
+                for key in _root_keys(item):
+                    yield key
+        elif "." in key:
+            yield key.split(".", 1)[0]
+        else:
+            yield key
+
+
+def parse_filter(filter):
+    """Parse a provided sequence of filters.
+
+    Parameters
+    ----------
+    filter : Sequence, Mapping, or str
+        A set of key, value tuples corresponding to a single filter. This
+        filter may itself be a compound filter containing and/or statements. The
+        filter may be provided as a sequence of tuples, a mapping-like object,
+        or a string. In the last case, the string will be parsed to generate a
+        valid set of filters.
+
+    Yields
+    ------
+    tuple
+        A key value pair to be used as a filter.
+
+    """
+    filter = list(filter)
+
+    if isinstance(filter, str):
```

Comment:
Will the filter ever be a `str` since we cast it as a list in line 301?

Reply:
Nope, good catch!